### ISSUE TITLE
Dynamic Range Compression Support

### ISSUE THREAD

**Issue by bobbens** [naev/naev#2062]
Currently we do a pretty bad job of dealing with lots of audio sources coming together, meaning large fights just saturate speakers/headphones and it sounds like trash. Having some sort of [dynamic range compression](https://en.wikipedia.org/wiki/Dynamic_range_compression) would be great. A possible reference implementation is https://github.com/velipso/sndfilter and it could likely be directly integrated as an OpenAL filter. Of course it may be easier to go with a more simple approach like decrease the volume based on the number of sources, but that might just make things too quiet if we don't consider the actual volume of the sources.

**Comment by onpon4** [comment]
I can't say I understand how dynamic range compression is supposed to help this issue, but why not just drop previously played instances of a sound effect so that no more than X of them play at a time (where X is some determined value between 1 and 5 or so)? Lots of games do this; it's very easy to do and while positional sound complicates matters somewhat, in practice it works very very well as long as it's handled correctly.

🕵️

**Comment by PhoenixRiver** [comment]
Would a volume limiter not work better? That way you'd not reduce the volume of specific sound effects too much...

Coming at this from another angle what about the possibility of splitting sound effects into "external" (i.e. emenating from outside the ship e.g. other ships' weapons fire and engines , alarms etc.) and internal (e.g. interface sounds, alarms etc.). I'm not sure where to put your own ship's weapons fire... I feel like this might be a can of worms involving different timbres for your own weapons and engine sounds from those of others... deeper  say since they are modulated by vibrating the ship your in... maybe they'd be a third channel?
These two groups could be handled differently with a potential limiter being introduced for external noises with a message to console indicating that for the concentration of the pilot external sounds have been muted...

This muting of external sounds would mean that the majority of sounds (at least in most situations where there are large nubmers of sounds) would be muted but in context and would avoid drowning out sounds the player might need to hear and are more important to them... as it would a pilot.

Incidentally if the player's ship's own sounds were in a third channel having different sounds effects wouldn't be necessary since a timbre of some sort could be applied to the default sounds as a whole to differentiate them.

Internal cabin sounds might also be cool at add to the derelict family mission but I'd be hesitant to add them without more control over their volume... do we have a way to deal with that? Maybe a hidden slider that we don't yet expose to the UI?

Talking of which, splitting into these channels would give us the potential to alter the volume, or have the player able to alter the volume, of these independantly (as well as "multiply up" the number of sound effects we have) which might be nice :)

**Comment by bobbens** [comment]
Sound dropping and limiting volume are susceptible to artefacts and don't really solve the issue. It may be good enough to make it not noticeable in some cases, but things like shotguns generating multiple sounds at once and at basically a same position would easily break it. Dynamic range compression is more work to implement but solves it all with no corner cases in a single elegant framework.

> Incidentally if the player's ship's own sounds were in a third channel having different sounds effects wouldn't be necessary since a timbre of some sort could be applied to the default sounds as a whole to differentiate them.

Player's ship sounds (mainly engine and things like activated outfits), are already separate to environment sounds.

> Internal cabin sounds might also be cool at add to the derelict family mission but I'd be hesitant to add them without more control over their volume... do we have a way to deal with that? Maybe a hidden slider that we don't yet expose to the UI?

I do think that could be more confusing and work to implement. I do like the idea of environment sounds when landing on different types of stations / planets though.

> Talking of which, splitting into these channels would give us the potential to alter the volume, or have the player able to alter the volume, of these independantly (as well as "multiply up" the number of sound effects we have) which might be nice :)

Somewhat related, you can modify the engine sound independently in nightly.

**Comment by PhoenixRiver** [comment]
Since apparently I don't know what has and hasn't already been implimented! ;) how aobut only applyin gthe dynamic sound compression to external to the ship sounds which are (presumably) less important than internal ship sounds and so the risk of overly quietening them is also less important? Or am I just repeating what you suggested... probably!

All these fun things to play with and ideas to try  but so little time to read code... at least I can proofread pretty much in my sleep! ;)

**Comment by bobbens** [comment]
> If "sound dropping" referring to what I'm talking about, um, no? Lots and lots of games do it with no ill effect. Dragon Ball Xenoverse 2, for example, does it with no ill effect: if it plays a "hit" sound, it cancels any previously playing "hit" sound, prioritizing the "hit" sounds closest to you. Et cetera, et cetera. I could refer to tons of other examples. It's _everywhere_.

I don't see how it being everywhere implies that it is the best solution. Heuristic-based approaches are a pain in the ass to maintain, and you literally have to think of all the edge cases to work. I can see a clear example where what you propose fails:

Beam attack coming close to player and nearby ship firing vulcan turret. The beam sound gets overwritten by the vulcan pellets (it shoots a lot), and when the pellets are long gone and the beam is on top of the player, it is no longer playing a sound. Sure you can solve this by considering sound length in the heuristic, but you're making it more complicated and there are still other corner cases left all over.

Dynamic range compression perfectly solves this in a single framework, without heuristics nor the need for a sound engineer. Only potential downside is implementation complexity, which doesn't matter if you're not the one implementing it.

**Comment by bobbens** [comment]
> Erm, that's not how you would solve this, how you would solve this is to prevent the problem from occurring in the first place by making beams use a different channel than bolts, or you could go further and make each individual sound have its own channel (literally the simplest possible implementation). Even the example I showed above does that (notice it has a laser/plasma channel, a ray/missile channel, and a dedicated hit sound channel). You're making this sound way more complicated than it actually is.

You're using hardcoded channels, which is something we can't really do in Naev given the amount and diversity of outfits. Furthermore that doesn't solve the problem. If you have too many groups, you have too much loudness, if you have too few, you get dropped ones. Finding a balance is non-trivial when you don't have a small amount of hardcoded sound sources, which is what happens in Naev.

> Dynamic range compression will also make the game sound worse, even if you do it perfectly, and having uniform loudness everywhere will make long game sessions exhausting. Canceling out the difference between quiet and loud would also, if I'm not mistaken, more or less eliminate any volume difference between near and distant (after all it's the direct audio stream you'd have to compress the dynamic range of, not individual sounds).

I don't actually think you understand dynamic range compression. It doesn't have to actually make loudness uniform, but can decrease loudness after a certain point (downward compression). When it's not triggered, the game will sound exactly the same. When it triggers you get a drop-off in total loudness. Apparently our biological system does something similar.

Finally, I really don't see why you're choosing this hill to die on. You're making arguments for making the code full of heuristics and much more harder to handle, and arguing against a sort of strawman characterization of dynamic range compression as you don't actually seem to understand what I am proposing. You should try to understand the wikipedia article I linked at the beginning before getting into a what I can only call a pointless argument. I'm not going to indulge you any further discussion on this topic, so please leave this issue for discussion on potential libraries and approaches for dynamic range compression.

**Comment by onpon4** [comment]
> You're using hardcoded channels, which is something we can't really do in Naev given the amount and diversity of outfits. Furthermore that doesn't solve the problem. If you have too many groups, you have too much loudness, if you have too few, you get dropped ones. Finding a balance is non-trivial when you don't have a small amount of hardcoded sound sources, which is what happens in Naev.

The most extreme case would be to dynamically assign a channel for each sound file. The slightly less extreme case would be to designate certain types of sound effects as "infinite-playing", and give every other sound file its own dedicated channel.

> I don't actually think you understand dynamic range compression. It doesn't have to actually make loudness uniform, but can decrease loudness after a certain point (downward compression). When it's not triggered, the game will sound exactly the same. When it triggers you get a drop-off in total loudness. Apparently our biological system does something similar.

You're right to say that I don't have a very good understanding of how dynamic range compression works, but the only way I can imagine dynamic range compression doing anything about duplicate sound files is capturing the mixed audio (after the audio mixer has already clipped and distorted audio above a volume it can handle) and putting it through some sort of dynamic range compression buffer to limit the volume to some "reasonable" level. That's only possible if you eliminate loud sounds above that threshold, meaning reducing the difference between quiet and loud will happen. If you have some other method of using dynamic range compression to prevent duplicated audio from attacking your eardrums, I'd love to hear it.

> Finally, I really don't see why you're choosing this hill to die on. You're making arguments for making the code full of heuristics and much more harder to handle, and arguing against a sort of strawman characterization of dynamic range compression as you don't actually seem to understand what I am proposing.

I can't exactly make a strawman of what you want to do when you haven't even explained what that is. Maybe consider that your proposal is unclear and I said at the very start that I didn't understand it? You don't have to jump to "strawman" every time I don't understand what you're talking about, and doing that is one of the most unprofessional things you do. To strawman is to be deliberately dishonest, it's not just a meaningless word to throw around randomly when you have a disagreement.

Also, again, what I'm proposing isn't for "making the code full of heuristics", and I don't know why you're fixating on the word "heuristic" as if it's some complicated thing. It's only *one* heuristic that's required, in one place, and the heuristic is just a simple "if" statement. The one Starfighter uses can literally be copied wholesale.

> You should try to understand the wikipedia article I linked at the beginning before getting into a what I can only call a pointless argument. I'm not going to indulge you any further discussion on this topic, so please leave this issue for discussion on potential libraries and approaches for dynamic range compression.

The Wikipedia article doesn't mention dynamic range compression as a means to prevent amplification of mixed sound effects from passing a threshold, so it's rather unfair to blame the fact that I have little idea what you're proposal is on whether or not I've read that Wikipedia article. All you've "proposed" is generically "use dynamic range compression", without any explanation of how that is supposed to work, and the only other thing you've done in this thread is shoot down alternative ideas for achieving the same goal in a much simpler fashion, including my idea which I went so far as to show an example implementation of.

**Comment by onpon4** [comment]
bobbens, what you are doing right now is bullying.

I don't need a graph showing what dynamic range compression does, which actually, no, is not clear; that you expect everyone who looks at that visual to immediately understand everything about dynamic range compression and never once make the slightest error in talking about it is ableist. I can't read diagrams like that, no. I'm autistic, and I don't process information like that visually. What I see there is a bent line, and it means nothing to me even though I've seen someone explain it.

But even if I could actually read that, that has nothing to do with why I don't understand your proposal, because what I don't understand is *how* you intend to use dynamic range compression to solve the problem at hand. You haven't explained it at all. I've done my absolute best to interpret what you are proposing in spite of you not explaining it, and you're accusing me of strawmanning without even explaining what I got wrong in spite of me requesting an explanation. I'm repeating myself, but strawmanning is not failing to understand a vague suggestion, it is deliberate dishonesty, and I really don't appreciate you bullying me with accusations of deliberate dishonesty just because I don't have a perfect understanding of how dynamic range compression works on a technical level and you're unwilling to explain how you're going to use dynamic range compression to prevent duplicate sound files from stacking and amplifying each other.

In the quote you claim is a "strawman" from your last post, the *only* thing I said that was inaccurate is "uniform loudness everywhere", and that's only if you take what I said absolutely literally. Lowering dynamic range *will* make the volume more uniform (that's what dynamic range compression means, regardless of whether you're boosting quiet sounds or turning down loud sounds). A more uniform volume *will* make the game sound worse, and it *will* make the game exhausting to listen to. The one thing I said which was clearly wrong, and which you quoted shortly before I edited, was the overstated "more or less eliminate any volume difference", which is why I edited that to "reduce volume difference" before I even saw your response. Neither of these mistakes are strawmen (neither my over-estimation of volume difference reduction due to limited understanding of dynamic range compression, nor my poorly worded summary of what dynamic range compression does). I have never in this entire conversation been deliberately dishonest.

You are also asserting your power in an incredibly abusive way right now by demanding I "refrain from non-productive discussion" because you claim my "insistence" is "toxic" while you have been attacking my character and intelligence. I have been nothing but polite in this thread in spite of your attacks (which started with you saying that you "don't actually think [I] understand dynamic range compression"). Being a community leader doesn't mean looking for excuses to shut down dissenting respectful views while you directly violate your own Code of Conduct by launching personal attacks.
